### ISSUE TITLE
[Proposal][Free function]

### ISSUE THREAD

**Issue by redradist** [dotnet/csharplang#2156]
This proposal is not original at all.
**Lets introduce to C# free functions !!**

**Why we should add free function to C# ?? We've already have static methods !?**
Static methods in class on the other hand is some-like a language **_hack_** ... Class was not designed to contain a lots of static methods. One of the main purposes of class is to hold a state of the instance.
Of course sometimes static methods is necessary !!
This feature will made C# language more clear, more expressive.

What do you think ? Share you opinion ...
Also everyone how like this proposal please mark as like it )

**Comment by YairHalberstadt** [comment]
> Class was not designed to contain a lots of static methods.

Given that static classes were designed by Anders Heijlsberg at the same time as he designed non-static classes, this is false. Both have existed since version 1.

Are static classes causing a problem now, that free functions will solve?

Saying bit will make the language more expressive is insufficient. Can you give a real life example, where allowing free functions would cause a substantial benefit to a developer?

By the way you may be interested in the `using static` directive. This brings a static class into scope so you can call it's methods as if they were free methods.

**Comment by stuartstein777** [comment]
Id quite like a function that I can put in a namespace and not have to put in a class.

It would reduce the number of nonsense classes that are just Helper.cs etc, which exist only to hold a function.

**Comment by gistofj** [comment]
@stuartstein777 you can always `using static MyNamespace.HelperClass` and just reference the `MyNamespace.HelperClass` static methods as if they were "free methods". :man_shrugging:

**Comment by HaloFour** [comment]
I want to say that top-level members has been discussed before but I can't find any open proposals on the subject.  It's probably over on [Roslyn](https://github.com/dotnet/roslyn) or buried in the [CodePlex](https://archive.codeplex.com/?p=roslyn) archive.

The question is what value would this really provide?  Sure, you can slightly reduce the amount of boilerplate when it comes to writing `static` classes of helper functions, but it doesn't really do anything else.  You can already reduce boilerplate at the callsite by using `using static` to import entire classes of `static` methods.  In my opinion, it wouldn't be worth the amount of effort that would need to go into it.

That said, seems there's a contingent on the C# design team that would like to unify the C# and C# Scripting languages.  If that happens not only would you be able to define top-level functions but also top-level variables and top-level code.  And I guess reference directives?  Who knows.

**Comment by redradist** [comment]
> 
> 
> I want to say that top-level members has been discussed before but I can't find any open proposals on the subject. It's probably over on [Roslyn](https://github.com/dotnet/roslyn) or buried in the [CodePlex](https://archive.codeplex.com/?p=roslyn) archive.
> 
> The question is what value would this really provide? Sure, you can slightly reduce the amount of boilerplate when it comes to writing `static` classes of helper functions, but it doesn't really do anything else. You can already reduce boilerplate at the callsite by using `using static` to import entire classes of `static` methods. In my opinion, it wouldn't be worth the amount of effort that would need to go into it.
> 
> That said, seems there's a contingent on the C# design team that would like to unify the C# and C# Scripting languages. If that happens not only would you be able to define top-level functions but also top-level variables and top-level code. And I guess reference directives? Who knows.

**Comment by HaloFour** [comment]
@redradist 

There is [C# Interactive](https://github.com/dotnet/roslyn/wiki/C%23-Interactive-Walkthrough) which can be used for that.

**Comment by redradist** [comment]
> 
> 
> @redradist
> 
> There is [C# Interactive](https://github.com/dotnet/roslyn/wiki/C%23-Interactive-Walkthrough) which can be used for that.

But C# Interactive should be also valid C# code. This will mean that I can share the same codebase between prototype and production (compiled code)

**Comment by HaloFour** [comment]
@redradist 

https://github.com/dotnet/roslyn/issues/14143#issuecomment-250795384

I don't recall seeing any further movement on that subject, though.

**Comment by jnm2** [comment]
https://github.com/dotnet/csharplang/issues/1180

**Comment by johnkellyoxford** [comment]
Interesting approach, liking and +1ing your own suggestion 🤔.

**Comment by YairHalberstadt** [comment]
> Interesting approach, liking and +1ing your own suggestion 🤔.

You can't deny it's effective 😉

**Comment by redradist** [comment]
> Interesting approach, liking and +1ing your own suggestion thinking.

Thanks =)

**Comment by Berrysoft** [comment]
C++/CLI supports this. It seems that the C++ compiler put the free functions into the internal class `<Module>`, thus these functions can't be seen outside the assembly.

**Comment by tonygiang** [comment]
I greatly support this feature for C#. Namespace is enough to avoid naming conflicts. Omitting the need for a wrapper class saves C# developers an extra level of indentation.

**Comment by redradist** [comment]
> 
> 
> I greatly support this feature for C#. Namespace is enough to avoid naming conflicts. Omitting the need for a wrapper class saves C# developers an extra level of indentation.

It would be nice if you like this propousal, to show Microsoft guys how many developers need this feature )

**Comment by tpetrina** [comment]
Problem with static classes as holders for static functions is that you _have_ to _name_ them. After a while you run out of good names. I mean, sometimes the good name for that static method(s) is the namespace itself.

**Comment by HaloFour** [comment]
@tpetrina 

> Problem with static classes as holders for static functions is that you _have_ to _name_ them.

Either you do, or the compiler does.  If these methods are accessible outside of the assembly that forces the language specification to define deterministic rules as to how that name is generated so that recompilation doesn't break existing consumers of those functions and works well for other languages in the ecosystem.  And the proposal to immediately follow this language feature would be more language features to explicitly name that auto-generated wrapper class, and to let the span files, if Kotlin is any indication.

Names are cheap and defining a static type is a burden of five tokens.

**Comment by tonygiang** [comment]
> 
> 
> Problem with static classes as holders for static functions is that you _have_ to _name_ them. After a while you run out of good names. I mean, sometimes the good name for that static method(s) is the namespace itself.

You could technically turn your static function wrapper class into a pseudo-namespace by marking it as a `partial` class. The problem is:
1. It feels like a hack (because it is a hack)
2. You get into troubles when you start to make partial wrapper classes behave like nested namespaces. Basically, you can only do a monolithic `Utils.Algorithm` class (name is just an example), you can't have more domain-specific helper namespaces like `Utils.Algorithm.Geometry`, `Utils.Algorithm.DateTime`, `Utils.Algorithm.Database`, etc... while still leaving common algorithms in `Utils.Algorithm`. The language designers have inadvertently promoted BDUF (Big Design Up Front) which is unfortunate.

**Comment by gulshan** [comment]
I think free functions encourages state-less function-centric designs- all required data goes into a function, get processed and gives a result. Where as OOP encourages stateful object centric designs. When combined with an affinity for immutability, function-centric a approach results into simpler and easier to manage programs with clear dataflow IMHO. And with extension methods, free functions can act like OOP methods. Rust uses extension method like feature to look like OOP. I use extension methods in my Kotlin solutions also. I think free function is needed for this purpose. Otherwise it does not enable anything which is not possible now.

**Comment by HaloFour** [comment]
@gulshan 

> Where as OOP encourages stateful object centric designs. When combined with an affinity for immutability, function-centric a approach results into simpler and easier to manage programs with clear dataflow IMHO.

You can accomplish all of the same with `static` classes, or even regular classes, just by exercising some discipline.  You need to exercise that same discipline with Kotlin, which lets you declare mutable top-level variables right alongside those top-level functions.  And Kotlin still generates a wrapper class around those top-level members as the JVM has no concept of "free functions", and Kotlin had to offer other language enhancements so that the developer could control the naming and organization of those top-level members.

**Comment by tpetrina** [comment]
> @tpetrina
> 
> > Problem with static classes as holders for static functions is that you _have_ to _name_ them.
> 
> Either you do, or the compiler does. If these methods are accessible outside of the assembly that forces the language specification to define deterministic rules as to how that name is generated so that recompilation doesn't break existing consumers of those functions and works well for other languages in the ecosystem. And the proposal to immediately follow this language feature would be more language features to explicitly name that auto-generated wrapper class, and to let the span files, if Kotlin is any indication.
> 
> Names are cheap and defining a static type is a burden of five tokens.

So make the compiler name those free functions as belonging to the `Assembly.Name` partial static class so that I don't have to create a unique static class name. And it should be a unique static class name since partial static classes are not a thing and non-static classes cannot house extension methods. And they falsely imply that there  _could_ be instances of that class.

Isn't that what compilers should do? Ensure sane defaults, pit of success, principle of least astonishment etc...

Since F# methods can already be called with `module.function` name from C#, I don't see this as an unsurmountable problem.

**Comment by CyrusNajmabadi** [comment]
i agree with @HaloFour.  Specifically "Names are cheap and defining a static type is a burden of five tokens."

This just doesn't seem like an issue worth investing effort into.

**Comment by HaloFour** [comment]
@tpetrina 

> And it should be a unique static class name since partial static classes are not a thing and non-static classes cannot house extension methods. And they falsely imply that there could be instances of that class.

The following is perfectly legal:

```cs
public static partial class Foo { }

public static partial class Foo { }
```

> Since F# methods can already be called with module.function name from C#, I don't see this as an unsurmountable problem.

But it is a problem, one that requires design and effort and invariably the want for more options and control.  And it's a problem that can be completely avoided by simply not trying to add a feature that is trying to pretend that static members of a class don't belong to a class.

**Comment by tpetrina** [comment]
> The following is perfectly legal:
> 
> ```cs
> public static partial class Foo { }
> 
> public static partial class Foo { }
> ```

Wow! How did I not know _that_? Thanks!
 
> But it is a problem, one that requires design and effort and invariably the want for more options and control. And it's a problem that can be completely avoided by simply not trying to add a feature that is trying to pretend that static members of a class don't belong to a class.

That argument works when you pretend that the _current_ design is alpha and omega. I would say that we shouldn't pretend functions belong to _any_ class, but you could argue back with "that is the case for the past 18 years". And you are right.

But semantically speaking we could add all global functions to module `$Unnamed` under the hood and call it a day. Otherwise I'll just have to resort to something like `public partial static class _Global_MyRootNamespaceWithoutDotsInIt_ {}`

**Comment by HaloFour** [comment]
@tpetrina 

> That argument works when you pretend that the current design is alpha and omega.

I'm arguing from the perspective of the CLR and the JVM, where all functions belong to a class.  Any language that offers "free" or "top-level" functions in either runtime does so by emitting some wrapper class and pretending that it doesn't exist.

> But semantically speaking we could add all global functions to module $Unnamed under the hood and call it a day.

Yes, you could, but that requires work, both to emit the type and to change how name resolution work so that these functions fall into scope.  And it will require further work if you intend for those functions to be public/exportable outside of that assembly.  IMO, unnecessary work, to avoid a name and five tokens.

I'd argue that the better angle for trying to get this into the language is aligning C# with CSX, although I don't even know if the latter is even a thing anymore.

**Comment by tpetrina** [comment]
You keep saying that it is five tokens but you are wrong. It is the _name_ that is the problem - which must follow some rules and then later on if you want to use it you must write `using static Full.Namespace.Path.To.Class`. And we know what is the hardest problem in computer science...

I am not opposed to the 4 tokens - can I get `public partial static {}`?

**Comment by johnkellyoxford** [comment]
> can I get `public partial static {}`

no
confusing and unnecessary

**Comment by HaloFour** [comment]
@tpetrina

> It is the name that is the problem

Clearly it's not if you think that the compiler should be in the business of using some arbitrary name.  Just pick a convention and stick to it, e.g. `Functions`.

> can I get `public partial static {}`?

I am a firm believer that if something is a part of a public surface that it should have an explicit name.

**Comment by redradist** [comment]
Guys, I think disscussion stuck in technical details ...

Anyway, free functions possible to implement

According my background (C, C++, Python, Java, C#, Javascript) it is up to library author to choose what is the best design choose ...
If you propose developer for its design to wrap all methods in classes, you just enforce developer to write boilerplate code, namespace is enoghouh to resolve names collisions, belive me, I know C++ ;) !!

In some cases it is better to use free functions, sometimes it is better to wrap methods in classes and express interface ...

My point is that it is up to domain task what is better for it !!

**Comment by HaloFour** [comment]
@redradist 

> Anyway, free functions possible to implement

Many things are possible to implement.  That doesn't mean that it is reasonable to do so.

> According my background (C, C++, Python, Java, C#, Javascript) it is up to library author to choose what is the best design choose ...

Every language that you've mentioned has its own grammar, syntax and structure limitations.  It's up to the library author to choose the language that fits the design.  It's not the responsibility of the language to accommodate all possible designs.  There are dozens of .NET languages, some of which support free functions.  You're free to use them.

Anyway, check out: https://github.com/dotnet/csharplang/issues/2765

**Comment by redradist** [comment]
> @redradist
> 
> > Anyway, free functions possible to implement
> 
> Many things are possible to implement. That doesn't mean that it is reasonable or desirable to do so.
> 
> > According my background (C, C++, Python, Java, C#, Javascript) it is up to library author to choose what is the best design choose ...
> 
> Every language that you've mentioned has its own grammar, syntax and structure limitations. It's up to the library author to choose the language that fits the design. It's not the responsibility of the language to accommodate all possible designs. There are dozens of .NET languages, some of which support free functions. You're free to use them.

I disagree with you, because language should help developers to express themselves, but not to stop doing something ;(

**Comment by CyrusNajmabadi** [comment]
> I disagree with you, because language should help developers to express themselves

Every language has limitations on what developers can express.  That's practically the definition of what a 'language' is in the first place :)

**Comment by theunrepentantgeek** [comment]
> I disagree with you, because language should help developers to express themselves

This is a very seductive statement - one that seems, on first encounter, like it simply must be true.

On reflection, however, I don't think it is.

Programming languages _constrain_ and _restrict_ as much (or more than) they _enable_ and _permit_.

C# opines that strings are immutable and can't be changed, only replaced, eliminating a whole class of potential defects. A library design that requires mutable strings is still possible, but it would be both painful to implement and difficult to use.

C# opines that classes requiring explicit cleanup should implement `IDisposable`, even going so far as to provide multiple forms of the **using** statement to make this straightforward. A library design that ignores this convention is still possible, but you lose assistance from the language and library features will be less discoverable by experienced developers.

C# requires that you declare the type of each parameter to a method, so that the compiler can work alongside each developer checking for problems. A library design where every parameter was declared as **object** or **dynamic** is possible, but ... yeah, I hope you see my point.

Every programming language is different - which is why each programming community has different idioms about _how things are done here_. Writing a C# library using C++ or JavaScript idioms won't give you a great result.

**Comment by redradist** [comment]
@CyrusNajmabadi 
> 
> 
> > I disagree with you, because language should help developers to express themselves
> 
> Every language has limitations on what developers can express. That's practically the definition of what a 'language' is in the first place :)

But it does not mean that Language should stuck with its basic design and do not reconsider itsself ...
If free functions proves need over the time demand than a nice language designer should add it if it does not break all language rules ;)

**Comment by tonygiang** [comment]
> Every language that you've mentioned has its own grammar, syntax and structure limitations. It's up to the library author to choose the language that fits the design. It's not the responsibility of the language to accommodate all possible designs. There are dozens of .NET languages, some of which support free functions. You're free to use them.

If this is the case, why evolve C# as a language as all? There should never have been a C# 2.0, 3.0, 4.0 or any version past 1.0. If you want generics, lambda expression, `var` type inference or optional parameters, just use other languages instead. It's not the responsibility of the language to accommodate all possible designs, right?

**Comment by CyrusNajmabadi** [comment]
> If this is the case, why evolve C# as a language as all? 

Your statement doesn't logically follow from @HaloFour's.  Halo is stating: C# doesn't need to be the language for everyone for all needs.  That doesn't mean that it cannot improve and become better for some users for some needs.

**Comment by redradist** [comment]
@CyrusNajmabadi 
> > If this is the case, why evolve C# as a language as all?
> 
> Your statement doesn't logically follow from @HaloFour's. Halo is stating: C# doesn't need to be the language for everyone for all needs. That doesn't mean that it cannot improve and become better for some users for some needs.

Unfortunately it goes against the official direction of .NET as platform ... =)
With .NET 5 Microsoft want to cover all platforms, it means that C# language will be used for all needs, it is contradiction to your statement ;)

**Comment by tpetrina** [comment]
@HaloFour 

> Clearly it's not if you think that the compiler should be in the business of using some arbitrary name. Just pick a convention and stick to it, e.g. `Functions`.

It is already doing that for anonymous types...
 
> I am a firm believer that if something is a part of a public surface that it should have an explicit name.

Of course it has an explicit name - it's literally the _function's name_. The only argument you have is that currently C# compiler requires that all functions belong to a class name. But for naming we have namespaces - their sole purpose is to provide a space for names.

@CyrusNajmabadi The thing is people in the wild whose first programming language don't see static methods of a static class as abnormal - they see global functions as a problem. Which is strange since it is literally the opposite that is the normal. Hence the wish to align C# with literally the every language in the world. You don't see static class/static function in lambda calculus, right?

It is obvious that there are some issues with the current implementation of C# compiler, but it is probably easier to implement free functions than async/await and pattern matching.

**Comment by redradist** [comment]
@tpetrina 
Could you please thumb up the proposal to show community and Microsoft in general that this feature is important ?

**Comment by HaloFour** [comment]
@redradist 

> With .NET 5 Microsoft want to cover all platforms, it means that C# language will be used for all needs,

No, it very much does not mean that.  It's not a goal of C# to make it behave like every other language that exists in order to appease every programmer out there.  MS makes several .NET languages and fosters an ecosystem of dozens more.  If you don't feel C# fits your needs, look elsewhere.  Nobody at Microsoft or on the C# team is worried about trying to win a popularity contest.

Features will be added based on what they bring to the language and whether or not they're deemed worthy of the massive effort that goes into adding them.  The **vast** majority of proposed features, to make the language "better" and more "expressive", will not be implemented.

@tpetrina 

> It is already doing that for anonymous types...

The names of anonymous types are not a part of a public API.  They are an implementation detail and an internal one, so the compiler is free to generate a random and unspeakable name.  The compiler does so in quite a few cases, but only where those names are implementation details that do not leak to the surface.  That's not the case here.  Any such name will have to be a part of the specification and the API contract.

> Of course it has an explicit name - it's literally the function's name. 

That's only half of the name.  The other half is the class name, which is a part of that contract.  The namespace is only part of the class name.

> Hence the wish to align C# with literally the every language in the world.

Very explicitly not a goal of the C# language.

> but it is probably easier to implement free functions than async/await and pattern matching.

Ease of implementation is far from the only factor that goes into deciding whether or not a feature should be added.  If the team added everything that was "easy" the language would quickly devolve into a mess.  Pattern matching and async/await bring real changes to the way developers can solve problems.  Free functions at best reduce a minor amount of boilerplate.

**Comment by tpetrina** [comment]
@HaloFour You keep repeating all this time that it is currently designed for each function to need a class name. We are merely challenging that assumption, isn't that normal?

Is there really any argument for having a static class as an owner of a function _besides_ the argument "well that is how it is currently the case"?

Static functions in a static class is a workaround, there should be no need to have this requirement.

**Comment by HaloFour** [comment]
@tpetrina 

> You keep repeating all this time that it is currently designed for each function to need a class name. We are merely challenging that assumption, isn't that normal?

Sure, you can challenge it, as others can challenge that challenge.

IIRC, CTS requires these functions to reside within classes.  The entirety of the .NET ecosystem expects it.  Namespaces don't exist except as a convention of the class name.  At best the compiler could emit a static class called `Functions` or similar to contain these "free" functions, but you can do the same with a minimum amount of effort.

> there should be no need to have this requirement.

Nor is there a reason to remove it.  There's no practical benefit to being able to write the shortest "Hello World" possible.

**Comment by tonygiang** [comment]
> > With .NET 5 Microsoft want to cover all platforms, it means that C# language will be used for all needs,
> 
> No, it very much does not mean that. It's not a goal of C# to make it behave like every other language that exists in order to appease every programmer out there. MS makes several .NET languages and fosters an ecosystem of dozens more. If you don't feel C# fits your needs, look elsewhere. Nobody at Microsoft or on the C# team is worried about trying to win a popularity contest.

The moment C# stops being designed to suit the need of its userbase, it becomes an opinionated language like Go. Many of us use C# because we fled from Go to begin with. Very similar conversations have happened in the Go community before: Look, if you use Go, you have to accept not having generics, you have to accept unused imports not being allowed, you have to accept not having inheritance, you have to accept enforced access levels through first-letter capitalization of function names (not through explicit `public` and `private` keywords). In other words: It's my way (in this case, the Go way) or the highway.

C# has not been known to be like that. The language has added templates from C++ (in the form of generics), lambda expression and Monads (in the form of LINQ and extension methods) from more functional languages, `var` type inference and `dynamic` type from more dynamic languages, `?.` null-safe operator from Groovy, etc... After all this time, it's too late to make C# into an opinionated language at this point imo.

**Comment by HaloFour** [comment]
@tonygiang 

C# has and will always be opinionated.  There are plenty of language proposals here that the language team has explicitly stated that they have no interest in implementing.  There will always be syntax not accepted and features which will never exist.  This is true of literally every programming language ever designed and that will ever be designed.

**Comment by tonygiang** [comment]
> @tonygiang
> 
> C# has and will always be opinionated. There are plenty of language proposals here that the language team has explicitly stated that they have no interest in implementing. There will always be syntax not accepted and features which will never exist. This is true of literally every programming language ever designed and that will ever be designed.

In which case, we come back to my earlier question: Why evolve C# as a language at all? The language team should just go make their own language instead of making anything past 1.0.

**Comment by HaloFour** [comment]
> In which case, we come back to my earlier question: Why evolve C# as a language at all? The language team should just go make their own language instead of making anything past 1.0.

It's not a binary proposition between accepting any/all changes or accepting no changes.  The language team will evolve C# based on what they feel benefits the users of the language enough to be worth the enormous cost in doing so.

As mentioned members of the language team has expressed some interest in [top-level members](https://github.com/dotnet/csharplang/issues/2765), at least where it may help merge the delta between C# and C# Script.

**Comment by redradist** [comment]
@HaloFour 
> 
> C# has and will always be opinionated. There are plenty of language proposals here that the language team has explicitly stated that they have no interest in implementing. There will always be syntax not accepted and features which will never exist. This is true of literally every programming language ever designed and that will ever be designed.

You are very stubborn !!
Not accepted, not, no way, never ...

Seems like you have to take day off and program on other languages which are also general purpose ;)

**Comment by jnm2** [comment]
There is a well-known class named `<Module>` emitted in every compilation. Members in this class are what you can access via `Module.GetMethods`/`GetFields`/etc. Since the class always exists, why not throw stuff in there?

**Comment by jnm2** [comment]
Ah, to answer my own question: methods and fields can't be namespaced that way.

**Comment by jnm2** [comment]
@redradist There's nothing about general purpose which requires free functions. This is cosmetic.

**Comment by redradist** [comment]
@HaloFour 
> > In which case, we come back to my earlier question: Why evolve C# as a language at all? The language team should just go make their own language instead of making anything past 1.0.
> 
> It's not a binary proposition between accepting any/all changes or accepting no changes. The language team will evolve C# based on what they feel benefits the users of the language enough to be worth the enormous cost in doing so.
> 
> As mentioned members of the language team has expressed some interest in [top-level members](https://github.com/dotnet/csharplang/issues/2765), at least where it may help merge the delta between C# and C# Script.

I like this statement:
@HaloFour 
>
> ... The language team will evolve C# based on what they feel benefits the users ...

Very strange how team will know what users DOES want ?

Maybe they need to ask developers ?
Or I see ... They read minds ... Okay, it explains everything !!

**Comment by HaloFour** [comment]
@redradist 

> Seems like you have to take day off and program on other languages which are also general purpose

I program in a wide variety of languages, including several that have top-level statements.  And every one of those languages that have top-level statements encourage the use of some kind of structure as soon as you're doing anything remotely non-trivial.

> Very strange how team will know what users DOES want ?

They do accept user feedback, through many channels including this one.  That doesn't mean that they consider all user feedback equally.  C# is not a democracy, the LDM is a benevolent dictatorship.

We're at an impasse.  We clearly disagree on whether or not a proposal like this is useful, and I doubt there's little either of us can do to change the opinion of the other.  That's perfectly fine, it's not like my opinion particularly matters.  But it is the onus of the proposer to try to persuade a member of the language team that it's worth implementing as a language feature.  Language features require a "champion" on the language design team to debate the design of the feature in language design meetings in order for them to have any chance of being implemented.  Good luck.

**Comment by gulshan** [comment]
I think C# language design process should be more structured(categorize feature requests), objective(what path C# wants follow, what makes it special from other languages), transparent(score feature requests) and procedure-oriented(stages of feature graduation). Is there any impact of GitHub votes? How else can an outside help to move forward a feature request? F# and Rust uses RFC model. There can be something similar here.

**Comment by HaloFour** [comment]
@gulshan 

The design process is documented here:

https://github.com/dotnet/csharplang#design-process

https://github.com/dotnet/csharplang/issues/1838
https://github.com/dotnet/csharplang/issues/1600

**Comment by gulshan** [comment]
Did not deny that, there is a design process.

**Comment by CyrusNajmabadi** [comment]
> In which case, we come back to my earlier question: Why evolve C# as a language at all? The language team should just go make their own language instead of making anything past 1.0.

There is a broad spectrum between:

> Why evolve C# as a language at all?

and that it's the 

> responsibility of the language to accommodate all possible designs

or that it needs to 

> cover all platforms

or 

> be used for all needs

Once again, the goal isn't to be the language for everyone for every purpose.  It's not going to adopt every feature that everyone wants.  The LDM will judiciously decide what it things are most valuable for the ecosystem and most healthy for the language moving forward.

**Comment by CyrusNajmabadi** [comment]
>  Is there any impact of GitHub votes? 

Yes.

> How else can an outside help to move forward a feature request? 

It can occasionally help a *small* amount if someone is willing to not only write up the proposal and language spec changes, but also do the impl.  But it is still dependent on the LDM wanting to take the feature.

> F# and Rust uses RFC model. There can be something similar here.

Please take up suggestions on how you would like design to work in another proposal.  Thanks!

**Comment by redradist** [comment]
@CyrusNajmabadi 
> 
> ... The LDM will judiciously decide what it things are most valuable for the ecosystem and most healthy for the language moving forward ...

Let's wait when team guys realizes that Kotlin beats them by "free" function and in C#10 we will back to discussion =)

**Comment by CyrusNajmabadi** [comment]
> Let's wait when team guys realizes that Kotlin beats them by "free" function and in C#10 we will back to discussion =)

I don't think you realize that the LDM doesn't operate by trying to beat out other languages.  Features are chosen if they're felt to actually be worth it.  There are thousands of potential features that can be added among hundreds of other languages out there.  It's not a competition to try to add them all.

**Comment by CyrusNajmabadi** [comment]
> Very strange how team will know what users DOES want ?

The LDM has many sources of feedback that are used to gauge what will be most valuable the ecosystem.

> Maybe they need to ask developers ?

"Asking" is one source.  There are other channels as well.

> Or I see ... They read minds ... Okay, it explains everything !!

I think you're mischaracterizing the situation.  The LDM can hear and determine that this is something some users want (and clearly some users also do *not* want).  They then evaluate if it's the right decision for the language itself.  

These are non-reasons for doing language changes.
1. some users want it.
2. other languages have it.
3. it would be easy to do it.

**Comment by CyrusNajmabadi** [comment]
> The moment C# stops being designed to suit the need of its userbase, it becomes an opinionated language like Go. 

C# has always been, and is likely to continue to be, an opinionated language.  I worked on helping design the language from 2.0 to 8.0, and we continually and consciously made strong opinions on hundreds to thousands of pieces of the language.  It's very much the language it is now precisely because the majority opinion of hte LDM was that this was the direction it was supposed to go in.

**Comment by CyrusNajmabadi** [comment]
> C# has not been known to be like that. The language has added templates from C++ (in the form of generics),

And it was opinionated that you were *not* getting templates.  That was intentional and specific.  And if someone came across and said "but i want templates.  look, C++ has them, and <some other language> is getting them soon!" that's not going to change things.

> lambda expression and Monads (in the form of LINQ and extension methods) from more functional languages, var type inference and dynamic type from more dynamic languages, ?. null-safe operator from Groovy, etc... All of which were features added after 1.0. After all this time, it's too late to make C# into an opinionated language at this point imo.

All of these were opinions.  It was opinions that these specific features were valuable and worth doing for the language.

At the same time, we also had opinions on hundreds (to thousands) of other features.  I remember numerous times during C#'s life that we evaluated and chose not to go forward with many potential proposals to add features from other languages into C#.  It's part and parcel of making C# what it is.  It's not a language that absorbs all features from other languages.  It's a language that has a core team driving the direction on what it thinks is best for the ecosystem.

**Comment by Grauenwolf** [comment]
> I'm arguing from the perspective of the CLR and the JVM, where all functions belong to a class.

That's not actually true. At the IL level, you can have functions that aren't part of the class.

I don't know of any language that actually does this. Nor can you call such functions from C#. But I have made working prototypes that use this feature.

**Comment by jrmoreno1** [comment]
@Grauenwolf : interesting.  Can you not call them from c#, or is there simply no way to directly get a reference to them?  Could you have something that c# would recognize return a delegate of some kind that refers to this free function?

**Comment by Grauenwolf** [comment]
Reflection doesn't offer a way to access it. Maybe if the defining assembly exposed it as a Delegate, but since I was writing the IL by hand I don't know how I would do that.

**Comment by YairHalberstadt** [comment]
Closing as championed in https://github.com/dotnet/csharplang/issues/3117

**Comment by redradist** [comment]
@CyrusNajmabadi @YairHalberstadt Any possibility to reconsider this feature ??
